parlai/tasks/igc/agents.py

```diff
@@ -0,0 +1,336 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Image Grounded Conversations (IGC) Task.
+
+See https://www.aclweb.org/anthology/I17-1047/ for more details. One must download
+the data from https://www.microsoft.com/en-us/download/details.aspx?id=55324
+prior to using this teacher.
+
+The images are then downloaded from the urls specified in the data. Unfortunately,
+not all links are live, and thus some examples do not have valid images.
+
+As there is no training set, we manually split 90% of the validation set
+into train.
+"""
+import csv
+import os
+
+from abc import ABC, abstractmethod
+from PIL import Image
+from typing import List, Dict, Any
+
+from parlai.core.build_data import download_multiprocess
+from parlai.core.teachers import AbstractImageTeacher
+
+
+class IGCTeacher(AbstractImageTeacher):
+    """
+    Teacher for IGC Task.
+
+    See https://arxiv.org/abs/1701.08251 for more details
+    """
+
+    def __init__(
+        self: AbstractImageTeacher, opt: Dict[str, Any], shared: Dict[str, Any] = None
+    ):
+        self.blank_image_id = '0000'
+        super().__init__(opt, shared)
+        if shared is not None:
+            self.valid_image_ids = shared['valid_image_ids']
+        if self.image_features_dict is not None:
+            self.image_features_dict[self.blank_image_id] = self.blank_image_features
+        self.multi_ref = opt.get('igc_multi_ref', False)
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Include arg.
+
+        for multi-reference labels.
+        """
+        AbstractImageTeacher.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('IGC Arguments')
+        agent.add_argument(
+            '--igc-multi-ref',
+            type='bool',
+            default=False,
+            help='specify to evaluate on multi-reference labels',
+        )
+
+    def image_id_to_image_path(self, image_id: str) -> str:
+        """
+        Return image path given image id.
+
+        As this function is used in building the image features, and some of the
+
+        :param image_id:
+            image_id key, for IGC this is a str
+
+        :return:
+            the image path associated with the given image key
+        """
+        if image_id not in self.valid_image_ids:
+            image_id = self.blank_image_id
+        return os.path.join(self.get_image_path(self.opt), image_id)
+
+    def get_data_path(self, opt: Dict[str, Any]) -> str:
```

Comment:
another opt

Reply:
also, should this one be private?

Reply:
this function, along with several others, are public methods in the parent `AbstractImageTeacher` class - normally i'd say yes, probably should make private, however these methods are encouraged to be overridden, so not sure what the correct approach is